qutebrowser/components/braveadblock.py

```diff
@@ -0,0 +1,319 @@
+    """Show missing dependency warning, if appropriate.
+
+    If the adblocking method is configured such that the Brave adblocker
+    should be used, but the optional dependency is not satisfied, we show an
+    error message.
+    """
+    method = config.val.content.blocking.method
+    if method in ("adblock", "both"):
+        message.warning(
+            "Ad blocking method is set to '{}' but 'adblock' dependency is"
+            " not installed.".format(method)
+        )
+
+
+_RESOURCE_TYPE_STRINGS = {
+    ResourceType.main_frame: "main_frame",
+    ResourceType.sub_frame: "sub_frame",
+    ResourceType.stylesheet: "stylesheet",
+    ResourceType.script: "script",
+    ResourceType.image: "image",
+    ResourceType.font_resource: "font",
+    ResourceType.sub_resource: "sub_frame",
+    ResourceType.object: "object",
+    ResourceType.media: "media",
+    ResourceType.worker: "other",
+    ResourceType.shared_worker: "other",
+    ResourceType.prefetch: "other",
+    ResourceType.favicon: "image",
+    ResourceType.xhr: "xhr",
+    ResourceType.ping: "ping",
+    ResourceType.service_worker: "other",
+    ResourceType.csp_report: "csp_report",
+    ResourceType.plugin_resource: "other",
+    ResourceType.preload_main_frame: "other",
+    ResourceType.preload_sub_frame: "other",
+    ResourceType.unknown: "other",
+    None: "",
+}
+
+
+def resource_type_to_string(resource_type: typing.Optional[ResourceType]) -> str:
+    return _RESOURCE_TYPE_STRINGS.get(resource_type, "other")
+
+
+class BraveAdBlocker:
+
+    """Manage blocked hosts based from /etc/hosts-like files.
+
+    Attributes:
+        enabled: Should we block ads or not
+        _has_basedir: Whether a custom --basedir is set.
+        _cache_path: The path of the adblock engine cache file
+        _in_progress: The DownloadItems which are currently downloading.
+        _done_count: How many files have been read successfully.
+        _finished_registering_downloads:
+            Used to make sure that if all the downloads finish really quickly,
+            before all of the block-lists have been added to the download
+            queue, we don't call `_on_lists_downloaded`.
+        _wip_filter_set:
+            If we're in midst of updating the block lists, this attribute
+            contains a "work-in-progress" filter set that will later be used to
+            create a new instance of `_engine`. Otherwise it's `None`.
+        _engine: Brave ad-blocking engine.
+    """
+
+    def __init__(self, *, data_dir: pathlib.Path, has_basedir: bool = False) -> None:
+        self.enabled = _should_be_used()
+        self._has_basedir = has_basedir
+        self._cache_path = data_dir / "adblock-cache.dat"
+        self._in_progress = []  # type: typing.List[downloads.TempDownload]
+        self._done_count = 0
+        self._finished_registering_downloads = False
+        self._wip_filter_set: typing.Optional[adblock.FilterSet] = None
```

Comment:
This [breaks](https://travis-ci.org/github/qutebrowser/qutebrowser/builds/722528177) Python 3.5 - while this will also be dropped very soon, I'd rather do that separately from this PR, so replace this by:

```suggestion
        self._wip_filter_set = None  # type: typing.Optional[adblock.FilterSet]
```